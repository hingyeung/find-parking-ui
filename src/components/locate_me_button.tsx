import React from 'react';
import { ApplicationState } from '../types';
import { connect } from "react-redux";
import { fetchCurrentLocationAndAvailableParkingsWithThunk } from '../actions';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import Fab from '@material-ui/core/Fab';
import MyLocationIcon from '@material-ui/icons/MyLocation';
import { ViewState } from 'react-map-gl';


type LocateMeButtonProps = {
  fetchCurrentLocation: (mapViewState: ViewState) => void;
  mapViewState: ViewState;
}

const LocateMeButton: React.FunctionComponent<LocateMeButtonProps & React.HTMLProps<HTMLButtonElement>> = (props) => {
  return (
    <Fab color="primary" aria-label="Locate me" onClick={(e) => props.fetchCurrentLocation(props.mapViewState)}>
      <MyLocationIcon/>
    </Fab>
  )
};

const mapStateToProps = (state: ApplicationState) => {
  return {mapViewState: state.mapViewState};
};

const mapDispatchToProps = (dispatch: ThunkDispatch<ApplicationState, void, AnyAction>) => ({
  // https://github.com/reduxjs/redux/issues/1676#issuecomment-215413478
  fetchCurrentLocation: () => dispatch(fetchCurrentLocationAndAvailableParkingsWithThunk())
});

export default connect(mapStateToProps, mapDispatchToProps)(LocateMeButton);