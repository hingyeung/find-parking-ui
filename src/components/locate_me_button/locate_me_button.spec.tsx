import React from "react";
import { create } from "react-test-renderer";
import { LocateMeButton } from './locate_me_button';
import { ViewState } from "react-map-gl";
import Fab from '@material-ui/core/Fab';

describe("LocateMeButton component", () => {
    let store,
        mockMapViewState: ViewState,
        mockFetchCurrentLocation: (viewState: ViewState) => void;

    beforeEach(() => {
        mockMapViewState = {
            latitude: 0,
            longitude: 0,
            zoom: 1
        };
        mockFetchCurrentLocation = jest.fn((viewState: ViewState) => { });
    });

    test("Matches the snapshot", () => {
        const mockFetchCurrentLocation = (viewState: ViewState) => { };
        const button = create(<LocateMeButton mapViewState={mockMapViewState} fetchCurrentLocation={mockFetchCurrentLocation} />);
        expect(button.toJSON()).toMatchSnapshot();
    });

    it("should call fetchCurrentLocation prop when clicked", () => {
        // const mockFetchCurrentLocation = jest.fn((viewState: ViewState) => {});
        const component = create(<LocateMeButton mapViewState={mockMapViewState} fetchCurrentLocation={mockFetchCurrentLocation} />);
        const instance = component.root;
        const button = instance.findByType(Fab);
        button.props.onClick();
        expect(mockFetchCurrentLocation.mock.calls.length).toBe(1);
    });
});