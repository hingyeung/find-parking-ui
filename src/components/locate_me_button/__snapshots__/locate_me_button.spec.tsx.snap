// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocateMeButton component Matches the snapshot 1`] = `
<button
  aria-label="Locate me"
  className="MuiButtonBase-root-11 MuiFab-root-1 MuiFab-primary-3"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiFab-label-2"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-14"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 8c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm8.94 3c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
      />
    </svg>
  </span>
  <span
    className="MuiTouchRipple-root-23"
  />
</button>
`;
